Sources/SwiftParser/Statements.swift

```diff
@@ -1038,4 +1051,63 @@ extension Parser.Lookahead {
     } while lookahead.at(.poundIf, .poundElseif, .poundElse) && lookahead.hasProgressed(&loopProgress)
     return lookahead.atStartOfSwitchCase()
   }
+
+  /// Returns `true` if the current token represents the start of an `if` or `while` statement body.
+  mutating func atStartOfStatementBody() -> Bool {
+    guard at(.leftBrace) else {
+      // Statement bodies always start with a '{'. If there is no '{', we can't be at the statement body.
+      return false
+    }
+    skipSingle()
+    if self.at(.endOfFile) {
+      // There's nothing else in the source file that could be the statement body, so this must be it.
+      return true
+    }
+    if self.at(.semicolon) {
+      // We can't have a semicolon between the condition and the statement body, so this must be the statement body.
+      return true
+    }
+    if self.at(.keyword(.else)) {
+      // If the current token is an `else` keyword, this must be the statement body of an `if` statement since conditions can't be followed by `else`.
+      // Note that `guard` condition does not use `atStartOfStatementBody`.
+      return true
+    }
+    if self.at(.rightBrace, .rightParen) {
+      // A right brace or parenthesis cannot start a statement body, nor can the condition list continue afterwards. So, this must be the statement body.
+      // This covers cases like `if true, { if true, { } }` or `( if true, { print(0) } )`. While the latter is not valid code, it improves diagnostics.
+      return true
+    }
+    if self.atStartOfLine {
+      // If the current token is at the start of a line, it is most likely a statement body. The only exceptions are:
+      if withLookahead({ $0.atStartOfStatementBody() }) {
+        // If a statement body starts after this on, we actually have a closure followed by the statement body e.g.
+        // if true, { true }
+        // {
+        //   print("body")
+        // }
+        return false
```

Comment:
Sorry if I missed the previous discussion, but do we really care this case? Closure alone cannot be a condition. it must be used/called some way (e.g. call it, or make it a part of a sequence expression). This example case is ambiguous, it can be missing `()` after `{ true }` or missing `()` after `{ print("body") }`. I don't think it's worth to lookahead again if there's any strong reason.

Reply:
I agree that's a odd case  but today that's a valid and the parser reports a `Function produces expected type 'Bool'; did you mean to call it with '()'?` error so without this lookahead there would be a regression.

Reply:
That's not a valid condition today, it's just parsed as a condition expression and diagnosed in type checker. Since it's invalid anyway, I wouldn't consider it's a regression even if we don't parse it as a closure.

Reply:
I don't feel comfortable changing existing behavior and as a user I consider the type checker diagnostic more helpful because is probably more close to the user intention.

Reply:
I can add `at(.leftBrace)` before the `withLookahead` to avoid actually looking ahead if not necessary.

Reply:
But that's _before_ the trailing comma support, so there was no ambiguity. Supporting trailing comma introduces some ambiguity, and for example, I believe the user intension is that `{ return true }` is the body in the following case:
```swift
func test() -> Bool {
  if
    condition1,
    condition2,
  {
    return true
  }
  
  { print("something") }
}
```

Reply:
Sorry, I took I time from this PR and missed your last answer. That's a really good counterexample. Honestly, I'm good with either behavior, whatever is less controversy. Could you weight in @ahoppen?

Reply:
I agree with @rintaro.

Reply:
I will change then and also resolve the conflicts with main branch.